simple/msg_data_integrity.c

```diff
@@ -0,0 +1,717 @@
+        } else {
+                msg_index = 0;
+        }
+
+	expected_data = (char *) malloc(size);
+        if (!expected_data) {
+                perror("malloc");
+                return -1;
+        }
+
+        for (j = 0; j < size; j++) {
+                expected_data[j] = msg_char[msg_index++];
+                if (msg_index >= msg_char_length) {
+                        msg_index = 0;
+                }
+        }
+
+	do {
+		ret = fi_cq_read(rcq, &comp, 1);
+		if (ret < 0 && ret != -FI_EAGAIN) {
+			if (ret == -FI_EAVAIL) {
+				cq_readerr(rcq, "rcq");
+			} else {
+				FT_PRINTERR("fi_cq_read", ret);
+			}
+			return ret;
+		}
+	} while (ret == -FI_EAGAIN);
+
+	recv_data = (char *)recv_buf;
+
+	for(j = 0; j < size; j++) {
+		if (expected_data[j] == recv_data[j]) {
+			continue;
+		}
+		first_error_byte = j;
+		break;
+	}
+	if (first_error_byte == -1) {
+		goto no_error;
+	}
+	if (iter < 0) {
+		printf("Test Name = %s, Sync test. Size = %d First_Error_byte = %d\n", 
+			test_name, size, first_error_byte);
+	} else {
+		printf("Test Name = %s, Iteration = %d Size = %d"
+			" First_Error_byte = %d\n", test_name, iter,
+			size, first_error_byte);
+	}
+	first_error_byte = -1;
+
+	if (allowed_error_count > 0) {
+		error_count++;
+	}
+	if (detailed_error) {
+                print_error = 1;
+        }
+
+no_error:
+	if (detailed_output || print_error) {
+		print_byte_array("Expd Buffer", expected_data, size);
+		print_byte_array("Recv Buffer", recv_data, size);
+	}
+
+	memset(recv_buf, 0, buffer_size);
+	if (iter < 0) {
+		next_size = transfer_size;
+	}
+	ret = fi_recv(ep, recv_buf, next_size, fi_mr_desc(recv_mr), 0, recv_buf);
+        if (ret)
+                FT_PRINTERR("fi_recv", ret);
+
+	free(expected_data);
+	if (error_count > allowed_error_count) {
+		printf("More than %d Corrupt packets found\n",
+			allowed_error_count);
+		ret = 1;
+	}
+	return ret;
+}
+
+static int sync_test(void)
+{
+	int ret;
+
+	ret = wait_for_completion(scq, max_credits - credits);
+	if (ret) {
+		return ret;
+	}
+	credits = max_credits;
+
+	ret = opts.dst_addr ? send_xfer(16, -1) : recv_xfer(16, -1);
+	if (ret) {
+		return ret;
+	}
+	return opts.dst_addr ? recv_xfer(16, -1) : send_xfer(16, -1);
+}
+
+static int run_test()
+{
+	int ret, i;
+
+	transfer_size = opts.transfer_size;
+	ret = sync_test();
+	if (ret)
+		return ret;
+
+	clock_gettime(CLOCK_MONOTONIC, &start);
+	for (i = 0; i < opts.iterations; i++) {
+		ret = opts.dst_addr ? send_xfer(opts.transfer_size, i) :
+				 recv_xfer(opts.transfer_size, i);
+		if (ret)
+			return ret;
+
+		ret = opts.dst_addr ? recv_xfer(opts.transfer_size, i) :
+				 send_xfer(opts.transfer_size, i);
+		if (ret)
+			return ret;
+	}
+	clock_gettime(CLOCK_MONOTONIC, &end);
+
+	if (opts.machr)
+		show_perf_mr(opts.transfer_size, opts.iterations, &start, &end, 2, opts.argc, opts.argv);
+	else
+		show_perf(test_name, opts.transfer_size, opts.iterations, &start, &end, 2);
```

Comment:
Do we need to `show_perf` or `show_perf_mr` at all?  The performance of this test is basically irrelevant.  It would be nice to emit some sort of output for each message size tested, but the performance numbers don't seem like the right thing to show.  Maybe just one line for each message size, indicating `   128... OK` or something like that.